### ISSUE TITLE
Maintain documentation in a machine readable format

### ISSUE THREAD

**Issue by lewis6991** [luvit/luv#632]
### Problem

Neovim currently [translates](https://github.com/neovim/neovim/blob/master/runtime/doc/luvref.txt) the markdown documentation from this repo into its own vim-help format. This is currently done manually when we update Luv.

Further to this https://github.com/folke/neodev.nvim parses this documentation (in vim-help format) to generate [type annotations](https://github.com/folke/neodev.nvim/blob/main/types/stable/luv.lua) for luv (for use with [Lua-LSP](https://github.com/LuaLS/lua-language-server)).

### Request

To significantly simplify both of these workflows, can [docs.md](https://github.com/luvit/luv/blob/e8e7b7e13225348a8806118a3ea9e021383a9536/docs.md) be ported to a machine readable format (I propose) Lua, from which `docs.md` can be generated from a small auxiliary script?

I am happy to implement this myself. Just want to know in advance if this is an acceptable solution?

**Comment by squeek502** [comment]
Sounds good to me!

**Comment by Bilal2453** [comment]
So I have actually been working on this for a while now.  My idea was to [fully type Luvit out](https://github.com/Bilal2453/luvit-meta) (including [Luv](https://github.com/Bilal2453/luvit-meta/blob/main/library/uv.lua)) in the EmmyLua annotation, then parse that into JSON, then from there we can use that JSON for whatever else (including regenerating the EmmyLua annotations); Kinda the bootstrapping approach, but I thought that would make the most sense for me.

Currently I just need to choose the JSON structure and write the parser for it, from there we can write a markdown generator that produces something similar to the current docs.  This will end up generating about the same docs we have with the exception of the introduction, since the JSON will only contain actual data types and functions with their descriptions.

The data includes: classes, methods, descriptions for methods and classes (identical to the one we have in docs), arguments, returns, returns that cannot be discarded[1], constants, and *a lot* more additional data structures that can be returned by Libuv (like for example, the exact possible returns of `req_get_type`), or accepted by it (like the exact protocols accepted by `uv.socketpair`).

**Comment by lewis6991** [comment]
My idea was to not write a parser at all (and hence no dependency) and just keep the structure in source in Lua. From that we can render the annotations and documentation.

If we want to go to the parser root then I'd probably just implement a treesitter parser since that would be more useful in different projects.

**Comment by Bilal2453** [comment]
That is an interesting idea.  So basically, use the Lua structure that hold the docs to generate the markdown for the docs?

**Comment by lewis6991** [comment]
> That is an interesting idea.  So basically, use the Lua structure that hold the docs to generate the markdown for the docs?
> 

Yep. Basically keep the source in the simplest and easiest to process format, from which we render it into more complex formats.

**Comment by Bilal2453** [comment]
I was actually going the LPeg parser route here.  And the end result would be a JSON that then could be parsed into whatever.  But having this in Lua directly would indeed remove the JSON dependency which would be neat.

Was there anything preventing you from already doing that?  I mean you got the annotation thing for the LLS for a while now.

I guess I will instead be parsing the LLS annotations into Lua structures then, shouldn't be that more hard compared to JSON I believe.   My end goal was effectively re-documenting Luvit, which already has its docs in JSON but it depends on Nodejs, I preferred redoing it all over working with that lol.

**Comment by lewis6991** [comment]
When I raised the issue I wasn't aware the annotations we're already in LLS so that reduced the priority of this. However we still want this so we can render the docs in vimhelp format.

**Comment by SinisterRectus** [comment]
I don't care what the format is as long as it's eventually human readable with minimal effort. A Lua schema sounds fine to me.

**Comment by Bilal2453** [comment]
> When I raised the issue I wasn't aware the annotations we're already in LLS so that reduced the priority of this. However we still want this so we can render the docs in vimhelp format.

I see, that sounds good to me then.  The way I am imagining this currently is we have a `docs` directory which holds on `data.lua`, this files simply returns a table that hold all the details one need to generate the docs, and a `generate-md.lua` that makes use of `data.lua` to generate the final markdown docs.  The final `docs.md` can reside there too, perhaps in a `results` sub-directory.

This way people can easily make other scripts to generate different types of output.  Also makes it easy to provide a script that scrapes Libuv docs for any updates saving manual work (since the docs were originally generated from there it seems)

**Comment by Bilal2453** [comment]
What Lua structure/format do you have in mind for this?

**Comment by lewis6991** [comment]
I haven't put any thought into the structure yet, but basically the bare minimum to generate what we need: annotations+markdown.

**Comment by Bilal2453** [comment]
Progress update: Just started *actually* working on this, will try to get something bootstrapped over at https://github.com/Bilal2453/luv-docgen, once that is done will be contributing something like `docs/data.lua` and a `docs/generate-markdown.lua` over here.

**Comment by clason** [comment]
Just linking this new tool here that could be used to generate the Markdown docs from annotations: https://github.com/CppCXY/emmylua-analyzer-rust/tree/main/crates/emmylua_doc_cli
